quinn-proto/src/endpoint.rs

```diff
@@ -537,18 +537,27 @@ where
         let (retry_src_cid, orig_dst_cid) = if server_config.use_stateless_retry {
             if token.is_empty() {
                 // First Initial
+                let mut random_bytes = vec![0u8; RetryToken::RANDOM_BYTES_LEN];
+                self.rng.fill_bytes(&mut random_bytes);
+
                 let token = RetryToken {
                     orig_dst_cid: dst_cid,
                     issued: SystemTime::now(),
-                }
-                .encode(&*server_config.token_key, &remote, &temp_loc_cid);
-                let mut buf = Vec::new();
+                    random_bytes: &random_bytes,
+                };
+
+                // Derive AeadKey
+                let aead_key = S::AeadKey::from_hkdf(&*server_config.token_key, token.random_bytes);
+
+                let encoded_token = token.encode(&aead_key, &remote, &temp_loc_cid);
```

Comment:
Can we move the key derivation inside `RetryToken::encode` to help keep things encapsulated?